Add support for ID3v2.4 multi-value tags

OxygenCobalt opened the pull request:
#### The problem

My app uses ExoPlayer's metadata parser to work around issues with `MediaStore`'s metadata handling. However, recent developments have required me to start handling multi-value tags, as many users want my app to support multiple artist/genre entries.

While the vorbis extractor supports multiple instances of the tag, the ID3v2 extractor does not, even though ID3v2.4 specifies that text information frames can have multiple values in them if delimited by a null terminator. This worsens the user experience with multi-value tags.

#### What this PR does

This PR deprecates the `value` field in `TextInformationFrame` and replaces it with a new array field called `values`. `Id3Decoder` populates this field with multiple values if it finds multiple tags delimited by null terminators. The now-deprecated `value` field is populated with the first item in `values`, to retain compatibility with the old implementation that would parse the first value only. 

This new algorithm should be backwards compatible with ID3v2.3 and ID3v2.2, as they only leverage null terminators as an informal shortcut to make serializing to C strings easier.

#### Further notes

I've modified other parts of the codebase to use the non-deprecated value to the most backwards-compatible extent possible. I have also added tests into `Id3Decoder` to handle the new functionality.

OxygenCobalt committed:
Add support for ID3v2.4 multi-value tags

Add support for multi-value tags based on null terminators.

These are specific to ID3v2.4, but is backwards compatible with
ID3v2.3, so no version checks are needed.

OxygenCobalt committed:
Clarify comments

Clarify that null-termianted multi-value separators are specific to
ID3v2.4.

OxygenCobalt committed:
Rename eos terminology

Rename indexOfEos to indexOfTerminator to better reflect how a
terminator != the end of a frame now.

OxygenCobalt committed:
Fix second-order issues

Fix second-order issues stemming from the addition of multi-value tags.

OxygenCobalt committed:
Merge branch 'multi-value-id3v2' into dev-v2

OxygenCobalt committed:
Add ID3v2 multi-value tests

Add tests for ID3v2 multi-value text frames.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java:
```diff
@@ -455,14 +455,29 @@ private static TextInformationFrame decodeTxxxFrame(ParsableByteArray id3Data, i
 
+    // In ID3v2.4, text information frames can contain multiple values delimited by a null
+    // terminator. Thus, we after each "end of stream" marker we actually need to keep looking
+    // for more data, at least until the index is equal to the data length.
```
I think you can remove this comment

OxygenCobalt replied:
Resolved in 0eb56f6.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java:
```diff
@@ -455,14 +455,29 @@ private static TextInformationFrame decodeTxxxFrame(ParsableByteArray id3Data, i
+    // In ID3v2.4, text information frames can contain multiple values delimited by a null
+    // terminator. Thus, we after each "end of stream" marker we actually need to keep looking
+    // for more data, at least until the index is equal to the data length.
+    ArrayList<String> values = new ArrayList<>();
```
nit: move this down a few lines closer to where it's first used (probably L470)

OxygenCobalt replied:
Resolved in 0eb56f6.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
 
-  public TextInformationFrame(String id, @Nullable String description, String value) {
+  @NonNull
+  public final String[] values;
```
Let's use `ImmutableList<String>` here instead - since arrays are unfortunately mutable

https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/collect/ImmutableList.html

OxygenCobalt replied:
Resolved in 157f1f0.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java:
```diff
@@ -479,10 +494,26 @@ private static TextInformationFrame decodeTextInformationFrame(
+
+      valueStartIndex = valueEndIndex + delimiterLength(encoding);
+      valueEndIndex = indexOfTerminator(data, valueStartIndex, encoding);
+    }
```
This looks very duplicated with the code above - would pulling this into a private method make sense?

OxygenCobalt replied:
Resolved in 0eb56f6.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java:
```diff
@@ -455,14 +455,29 @@ private static TextInformationFrame decodeTxxxFrame(ParsableByteArray id3Data, i
+
+    int valueEndIndex = indexOfTerminator(data, valueStartIndex, encoding);
+    while (valueStartIndex < valueEndIndex) {
+      String value = decodeStringIfValid(data, valueStartIndex, valueEndIndex, charset);
```
I don't think this will ever hit the 'not valid' branch of `decodeStringIfValid`, since you've asserted in the `while` that `valueStartIndex < valueEndIndex` and L467 and L477 ensure that `valueStartIndex < data.length` and `valueEndIndex <= data.length` respectively. So this can probably just be replaced with 

```
new String(data, valueStartIndex, valueEndIndex - valueStartIndex, charsetName)
```

OxygenCobalt replied:
Resolved in 0eb56f6.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
 
   @Nullable public final String description;
+
+  /** @deprecated Use {@code values} instead. */
```
Can you also mention that the first entry of `values` contains this value?

OxygenCobalt replied:
Resolved in 157f1f0.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
+    if (values.length > 0) {
+      this.value = values[0];
+    } else {
+      this.value = null;
```
I think this should `""` to be backwards compatible with the previous behaviour (please add a test for this)

OxygenCobalt replied:
Resolved in 157f1f0.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
 
 /** Text information ID3 frame. */
 public final class TextInformationFrame extends Id3Frame {
+  private final static String MULTI_VALUE_DELIMITER = ", ";
```
I would just in-line this, i think it makes the usages clearer

OxygenCobalt replied:
Resolved in 157f1f0.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
       case "TT2":
       case "TIT2":
-        builder.setTitle(value);
+        builder.setTitle(values[0]);
```
I think for this index-access to be safe you need to transform an empty array in the constructor to a single-entry array containing an empty string. Probably better to make this line (and similar lines below) robust in the case of an empty values list?

Please add a test for the case of passing in an empty values array.

OxygenCobalt replied:
You're right about this. In fact, it's actually more consistent with the rest of the metadata module to make `values == [""]` rather than make it an empty list. I'll make that change.

OxygenCobalt replied:
Resolved in 157f1f0.

icbaker reviewed (changes requested):
Thanks for the contribution! I added some initial thoughts

OxygenCobalt committed:
Rework ID3v2 text frame values interface

Rework the interface for ID3v2 text frames by:
- Switching from a mutable array to ImmutableList
- Never leaving the array empty, instead filling it with "" if there
are no values

This makes the surface safer and more in line with the rest of the
module.

OxygenCobalt committed:
Improve multi-value text frame parsing

Rework the parsing of multi-value text frames to reduce duplication
and increase efficiency.

OxygenCobalt committed:
Make text information frame values abstract

Make the values of TextInformationFrame an abstract list instead of an
explicit ImmutableList.

OxygenCobalt committed:
Fix non-updated constructor usage

Fix a constructor usage that tried to use an array instead of a list.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
 
-  public TextInformationFrame(String id, @Nullable String description, String value) {
+  @NonNull
+  public final List<String> values;
```
Can you add javadoc guaranteeing that this will always have at least one element?

Something like:

> The values of this ID3 frame. Will always have at least one element.

OxygenCobalt replied:
Resolved in 85c4d98.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,77 @@
     this.description = description;
-    this.value = value;
+    
+    if( values.size() > 0 ) {
```
Given that the spec says (emphasis mine):
> All ID3v2 frames consists of one frame header followed by **one or more** fields containing the actual information.

I think we should make this type match the spec, and reject an empty list with `IllegalArgumentException`

Then where we construct the list in `Id3Decoder` we can do the transform from 'no field' -> `[""]` before passing it in here, and it's part of the existing handling 'invalid' files logic (similar to how the current code uses `decodeStringIfValid` inside `Id3Decoder` instead of trying to do that transform in this constructor).

OxygenCobalt replied:
Resolved in 85c4d98.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java:
```diff
@@ -474,15 +475,38 @@ private static TextInformationFrame decodeTextInformationFrame(
+
+    // In ID3v2.4, text information frames can contain multiple values delimited by a null
+    // terminator. Thus, we after each "end of stream" marker we actually need to keep looking
+    // for more data, at least until the index is equal to the data length.
```
I think you can remove this comment too. IMO it only really makes sense as part of the 'diff' from the previous (current) behaviour, but that diff won't be visible to a reader in future (e.g. in 1 year). We should just assume that multiple values are 'natural' so no need to explicitly comment it.

OxygenCobalt replied:
Resolved in 85c4d98.

OxygenCobalt committed:
Move ID3v2 text frame fallback to ID3v2 decoder

Instead of the ID3v2 text frame falling back to a singleton list of an
empty string when the given values are empty, make that case throw
an exception within the text frame, and move that fallback behavior
into the ID3v2 decoder.

icbaker commented:
Looks good, thanks for all the changes! I'll get this merged.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,76 @@
   public void populateMediaMetadata(MediaMetadata.Builder builder) {
+    // Depending on the context this frame is in, we either take the first value of a multi-value
+    // frame because multiple values make no sense, or we join the values together with a comma
+    // when multiple values do make sense.
```
During internal review of this PR, @tonihei asked whether this split of "multiple values makes sense for field X but not field Y" is defined anywhere in the spec - or whether it's just your intuition?

I see that https://id3.org/id3v2.4.0-frames says:
> All text information frames supports multiple strings

Given that, we're wondering if it might be safer to just comma-separate every value when writing into a string field.

For `trackNumber` and other 'transform to a single int' fields we would probably still need to only use the first value, since there's no real way to join multiple values together.

What do you think?

(no need to make any additional changes to this PR, just reply to this comment thread and I'll make the changes to the internal version and they will get propagated back here after it's submitted internally)

OxygenCobalt replied:
It's largely intuition from my personal experience @icbaker. Feel free to make it more generalized if you believe it aligns with the specification better.

icbaker commented on library/extractor/src/main/java/com/google/android/exoplayer2/metadata/id3/TextInformationFrame.java:
```diff
@@ -19,52 +19,76 @@
   @Override
   public void populateMediaMetadata(MediaMetadata.Builder builder) {
+    // Depending on the context this frame is in, we either take the first value of a multi-value
+    // frame because multiple values make no sense, or we join the values together with a comma
```
It's been pointed out that these strings may end up directly in a UI (e.g. in a playback notification) and using a comma to separate a list of items is not appropriate for all locales.

We have an [`exo_item_list` resource definition](https://github.com/google/ExoPlayer/blob/r2.18.2/library/ui/src/main/res/values/strings.xml#L127) which gives a localised way to build up a list of items but it's currently defined in the UI module, so not available from here (the extractor module).

I understand you sent this PR because you need access to these multiple values. Do you need them to be automatically put in the `MediaMetadata`  object (i.e. handled in this method) or is it enough for them to be available from `TextInformationFrame` and your app can handle them from there? We're considering just keeping the existing behaviour of this method for now, due to the additional complexity of moving `exo_item_list` to e.g. the common module and also accessing it here (we'd need a `Context`). But if that makes this PR completely pointless for you then we should probably rethink that simplification :)

OxygenCobalt replied:
It's enough to just be available in `TextInformationFrame` @icbaker. Feel free to revert it back to it's original behavior. What I might propose is a refactor of MediaMetadata itself to handle multiple values, but that's something that can be done on your end.
